### ISSUE TITLE
Checkbox Performance

### ISSUE THREAD

**Issue by jamesmfriedman** [material-components/material-components-web-react#392]
There was a weird issue I came across where patching native getters and setters was causing problems with React events firing.
https://github.com/material-components/material-components-web/issues/4018

In exploring the differences in implementations, I came across some stats I think would be valuable to you. Short version, mat-react Checkbox re-renders itself constantly. This will definitely be a performance bottleneck on form heavy sites. I personally have a couple of apps that have a matrix of checkboxes, and this would crush the browser.

Checkbox Component mount
mat-react: 8 render cycles
rmwc: 2 render cycles

Checkbox Change
mat-react: 17 render cycles
rmwc: 2 render cycles

**Comment by moog16** [comment]
@jamesmfriedman thanks for reporting this...I read up on #4018 and is a concern. I'll add some notes there, and keep this issue open for investigating on the React side.

**Comment by jamesmfriedman** [comment]
it has to do with the gratuitous use of setState. A few things you could try.

- Pure component. I usually stay away from these because they require the caller to know what they're doing and maintain object identity
- Debounce setState. React has some setState pooling built in though so this isn't ideal.
- Don't maintain the checkbox state inside the component. Right now you have 3 sources of truth, the props, the internal state, and the checkboxes state
- before you call setState, check to see if the value has even changed. Right now you updated the checked state every time the component updates, causing a double render.

**Comment by gugu** [comment]
Do you use Checkbox with ripple or without? Removing ripple effect makes it much faster

**Comment by vargajacint** [comment]
@moog16 
The Checkbox component performance is really bad. I'm using a lot of checkbox, and i have a lot of performance issue, please fix, or just use PureComponent

**Comment by moog16** [comment]
Yup we haven't had time to get to this issue - I also have experienced this with the checkbox and have a few ideas on how to change it. Would you like to take the lead on this one?

**Comment by gugu** [comment]
You can use checkbox without ripple. Much faster

**Comment by gugu** [comment]
How to use checkbox without ripple:
`import { Checkbox } from '...'`; instead of `import Checkbox from '...';`
